File: Gui/Linn/Gui/Test/TestSceneGraph.cs
using Linn.TestFramework;
using Linn.Gui.Scenegraph;

namespace Linn {
namespace Gui {

internal class SuiteBasicNodeTests : Suite
{
    public SuiteBasicNodeTests() : base("Basic Node tests") {
    }
    
    public override void Test() {
        Node node1 = new Node();
        TEST(node1.Name == "Node1");
        TEST(node1.Namespace == "None");
        TEST(node1.Active == true);
        
        node1.Name = "TestNode";
        TEST(node1.Name == "TestNode");
        TEST(node1.Fullname == "None.TestNode");
        
        node1.Namespace = "TestNs";
        TEST(node1.Namespace == "TestNs");
        TEST(node1.Fullname == "TestNs.TestNode");
        
        Node node2 = new Node();
        TEST(node2.Name == "Node2");
        node2.Name = "TestNode2";
        node2.Namespace = "TestNs";
        TEST(node2.Name == "TestNode2");
        TEST(node2.Fullname == "TestNs.TestNode2");
        TEST(node2.Active == true);
    }
}

internal class SuiteTransformationTests : Suite
{
    public SuiteTransformationTests() : base("Transformation tests") {
    }
    
    public override void Test() {
        Node node1 = new Node();
        node1.Name = "Node1";
        Node node2 = new Node();
        node2.Name = "Node2";
        // set scenegraph structure
        // node1
        //    node2
        // check adding a node
        node1.AddChild(node2);
        TEST(node1.Parent == null);
        TEST(node2.Parent == node1);
        TEST(node1.Children.Count == 1);
        TEST(node2.Children.Count == 0);
        TEST(node1.Child(0) == node2);
        // check we cant add the node twice
        //TEST_THROWS(Linn.Gui.SceneGraph.Node.DuplicateElement, node1.AddChild, node2)
        
        // check setting active state
        node2.Active = false;
        TEST(node1.Active == true);
        TEST(node2.Active == false);
        
        node1.Active = false;
        TEST(node1.Active == false);
        TEST(node2.Active == false);
        
        node1.Active = true;
        TEST(node1.Active == true);
        TEST(node2.Active == true);
        
        // check removing a node
        node1.RemoveChild(node2);
        TEST(node2.Parent == null);
        TEST(node1.Children.Count == 0);
        // check we cant remove the node twice
        //TEST_THROWS(Linn.Gui.SceneGraph.Node.NodeNotParent, node1.RemoveChild, node2)
        node2.Parent = node1;
        //TEST_THROWS(Linn.Gui.SceneGraph.Node.ElementNotFound, node1.RemoveChild, node2)
        node2.Parent = null;
        
        // check initial translations
        Vector3d trans = node1.LocalSrt.Translation;
        TEST(trans.X == 0 && trans.Y == 0 && trans.Z == 0);
        trans = node1.WorldSrt.Translation;
        TEST(trans.X == 0 && trans.Y == 0 && trans.Z == 0);
        trans = node2.WorldSrt.Translation;
        TEST(trans.X == 0 && trans.Y == 0 && trans.Z == 0);
        
        // check setting new local translation
        Vector3d newTrans = new Vector3d(100, 50, 10);
        node1.LocalTranslation = newTrans;
        trans = node1.LocalSrt.Translation;
        TEST(trans.X == 100 && trans.Y == 50 && trans.Z == 10);
        trans = node1.WorldSrt.Translation;
        TEST(trans.X == 100 && trans.Y == 50 && trans.Z == 10);
        trans = node2.LocalSrt.Translation;
        TEST(trans.X == 0 && trans.Y == 0 && trans.Z == 0);
        trans = node2.WorldSrt.Translation;
        TEST(trans.X == 0 && trans.Y == 0 && trans.Z == 0);
        
        // check setting new world translation
        newTrans = new Vector3d(200, 10, 20);
        node2.WorldTranslation = newTrans;
        trans = node2.LocalSrt.Translation;
        TEST(trans.X == 200 && trans.Y == 10 && trans.Z == 20);
        trans = node2.WorldSrt.Translation;
        TEST(trans.X == 200 && trans.Y == 10 && trans.Z == 20);
        
        // add a node and check it obtains the correct local and world translations from its parent
        node1.AddChild(node2);
        trans = node1.LocalSrt.Translation;
        TEST(trans.X == 100 && trans.Y == 50 && trans.Z == 10);
        trans = node1.WorldSrt.Translation;
        TEST(trans.X == 100 && trans.Y == 50 && trans.Z == 10);
        trans = node2.LocalSrt.Translation;
        TEST(trans.X == 200 && trans.Y == 10 && trans.Z == 20);
        trans = node2.WorldSrt.Translation;
        TEST(trans.X == 300 && trans.Y == 60 && trans.Z == 30);
        
        // check child node can set its world translation and implicitly its local translation
        newTrans = new Vector3d(150, 100, 20);
        node2.WorldTranslation = newTrans;
        trans = node2.LocalSrt.Translation;
        TEST(trans.X == 50 && trans.Y == 50 && trans.Z == 10);
        trans = node2.WorldSrt.Translation;
        TEST(trans.X == 150 && trans.Y == 100 && trans.Z == 20);
        
        // scenegraph structure
        // node 1
        //    node 2
        //        node 3
        // check propagation of translations and addition of local translation
        Node node3 = new Node();
        node3.LocalTranslation = new Vector3d(1, 1, 1);
        node2.AddChild(node3);
        trans = node3.WorldSrt.Translation;
        TEST(trans.X == 151 && trans.Y == 101 && trans.Z == 21);
        TEST(node3.Parent == node2);
        node1.AddChild(node3);
        TEST(node3.Parent == node1);
        trans = node3.WorldSrt.Translation;
        TEST(trans.X == 101 && trans.Y == 51 && trans.Z == 11);
    }
}

public class TestProgram
{
    public static void Main() {
        new RendererNull();
        Runner runner = new Runner("SceneGraph tests");
        runner.Add(new SuiteBasicNodeTests());
        runner.Add(new SuiteTransformationTests());
        runner.Run();
    }
}

} // Gui
} // Linn

File: Public/Gui/Linn/Gui/Test/TestSceneGraph.py
#!/usr/bin/python
from Linn.TestFramework.TestFramework import *
import Linn.Gui.Resources.Layout
import Linn.Gui.SceneGraph.Node
import Linn.Gui.SceneGraph.Srt

class SuiteBasicNodeTests(Suite):
    def __init__(self):
        Suite.__init__(self, 'Basic Node tests')
    
    def Test(self):
        node1 = Linn.Gui.SceneGraph.Node.Node()
        TEST(node1.Name() == 'Node1')
        TEST(node1.Namespace() == None)
        TEST(node1.Active() == True)
        
        node1.SetName('TestNode')
        TEST(node1.Name() == 'TestNode')
        TEST(node1.Fullname() == 'None.TestNode')
        
        node1.SetNamespace('TestNs')
        TEST(node1.Namespace() == 'TestNs')
        TEST(node1.Fullname() == 'TestNs.TestNode')
        
        node2 = Linn.Gui.SceneGraph.Node.Node('TestNode2')
        node2.SetNamespace('TestNs')
        TEST(node2.Name() == 'TestNode2')
        TEST(node2.Fullname() == 'TestNs.TestNode2')
        TEST(node2.Active() == True)
        
class SuiteTransformationTests(Suite):
    def __init__(self):
        Suite.__init__(self, 'Transformation tests')
    
    def Test(self):
        node1 = Linn.Gui.SceneGraph.Node.Node('Node1')
        node2 = Linn.Gui.SceneGraph.Node.Node('Node2')
        # set scenegraph structure
        # node1
        #    node2
        # check adding a node
        node1.AddChild(node2)
        TEST(node1.Parent() == None)
        TEST(node2.Parent() == node1)
        TEST(node1.NumChildren() == 1)
        TEST(node2.NumChildren() == 0)
        TEST(node1.Child(0) == node2)
        # check we cant add the node twice
        TEST_THROWS(Linn.Gui.SceneGraph.Node.DuplicateElement, node1.AddChild, node2)
        
        # check setting active state
        node2.SetActive(False)
        TEST(node1.Active() == True)
        TEST(node2.Active() == False)
        
        node1.SetActive(False)
        TEST(node1.Active() == False)
        TEST(node2.Active() == False)
        
        node1.SetActive(True)
        TEST(node1.Active() == True)
        TEST(node2.Active() == True)
        
        # check removing a node
        node1.RemoveChild(node2)
        TEST(node2.Parent() == None)
        TEST(node1.NumChildren() == 0)
        # check we cant remove the node twice
        TEST_THROWS(Linn.Gui.SceneGraph.Node.NodeNotParent, node1.RemoveChild, node2)
        node2.iParent = node1
        TEST_THROWS(Linn.Gui.SceneGraph.Node.ElementNotFound, node1.RemoveChild, node2)
        node2.iParent = None
        
        # check initial translations
        trans = node1.LocalSrt().Translation()
        TEST(trans.iX == 0 and trans.iY == 0 and trans.iZ == 0)
        trans = node1.WorldSrt().Translation()
        TEST(trans.iX == 0 and trans.iY == 0 and trans.iZ == 0)
        trans = node2.WorldSrt().Translation()
        TEST(trans.iX == 0 and trans.iY == 0 and trans.iZ == 0)
        
        # check setting new local translation
        newTrans = Linn.Gui.SceneGraph.Srt.Vector3d(100, 50, 10)
        node1.SetLocalTranslation(newTrans)
        trans = node1.LocalSrt().Translation()
        TEST(trans.iX == 100 and trans.iY == 50 and trans.iZ == 10)
        trans = node1.WorldSrt().Translation()
        TEST(trans.iX == 100 and trans.iY == 50 and trans.iZ == 10)
        trans = node2.LocalSrt().Translation()
        TEST(trans.iX == 0 and trans.iY == 0 and trans.iZ == 0)
        trans = node2.WorldSrt().Translation()
        TEST(trans.iX == 0 and trans.iY == 0 and trans.iZ == 0)
        
        # check setting new world translation
        newTrans = Linn.Gui.SceneGraph.Srt.Vector3d(200, 10, 20)
        node2.SetWorldTranslation(newTrans)
        trans = node2.LocalSrt().Translation()
        TEST(trans.iX == 200 and trans.iY == 10 and trans.iZ == 20)
        trans = node2.WorldSrt().Translation()
        TEST(trans.iX == 200 and trans.iY == 10 and trans.iZ == 20)
        
        # add a node and check it obtains the correct local and world translations from its parent
        node1.AddChild(node2)
        trans = node1.LocalSrt().Translation()
        TEST(trans.iX == 100 and trans.iY == 50 and trans.iZ == 10)
        trans = node1.WorldSrt().Translation()
        TEST(trans.iX == 100 and trans.iY == 50 and trans.iZ == 10)
        trans = node2.LocalSrt().Translation()
        TEST(trans.iX == 200 and trans.iY == 10 and trans.iZ == 20)
        trans = node2.WorldSrt().Translation()
        TEST(trans.iX == 300 and trans.iY == 60 and trans.iZ == 30)
        
        # check child node can set its world translation and implicitly its local translation
        newTrans = Linn.Gui.SceneGraph.Srt.Vector3d(150, 100, 20)
        node2.SetWorldTranslation(newTrans)
        trans = node2.LocalSrt().Translation()
        TEST(trans.iX == 50 and trans.iY == 50 and trans.iZ == 10)
        trans = node2.WorldSrt().Translation()
        TEST(trans.iX == 150 and trans.iY == 100 and trans.iZ == 20)
        
        # scenegraph structure
        # node 1
        #    node 2
        #        node 3
        # check propagation of translations and addition of local translation
        node3 = Linn.Gui.SceneGraph.Node.Node('Node3')
        node3.SetLocalTranslation(Linn.Gui.SceneGraph.Srt.Vector3d(1, 1, 1))
        node2.AddChild(node3)
        trans = node3.WorldSrt().Translation()
        TEST(trans.iX == 151 and trans.iY == 101 and trans.iZ == 21)
        TEST(node3.Parent() == node2)
        node1.AddChild(node3)
        TEST(node3.Parent() == node1)
        trans = node3.WorldSrt().Translation()
        TEST(trans.iX == 101 and trans.iY == 51 and trans.iZ == 11)
        


gRunner = Runner('SceneGraph tests')
gRunner.Add(SuiteBasicNodeTests())
gRunner.Add(SuiteTransformationTests())
gRunner.Run()
